### DISCUSSION TITLE
Regulate robot speed - Robot too slow during the executions with EGM

### DISCUSSION THREAD

**Discussion by kappa95** [ros-industrial/abb_libegm#135] [General]
Hello community,

I started to use the the libraries with ROS Melodic and with another Windows PC with Robotstudio with the virtual controller set up with the Add-in StateMachine 1.0 (Robotware version 6.08.00). I am able to execute the examples of the repository [abb_robot_bringup_examples](https://github.com/ros-industrial/abb_robot_driver/tree/master/abb_robot_bringup_examples). The issue is that the robot is moving very slow, it takes "hours" to reach positions. 
I have tried also the examples found in this issue: [create code samples](https://github.com/ros-industrial/abb_libegm/issues/18#issuecomment-740013963) and executing the example a1 of the joint trajectory position in the following way:
Firstly execute from main the StateMachine programs on the robot controller
Terminal 1: `roscore`
Terminal 2: `rosrun libegm_samples a1_joint_trajectory_node`
Terminal 3: `rosservice call /rws/sm_addin/start_egm_joint "{}"`

**Comment by gavanderhoorn** [comment]
Have you changed the value of `\MaxSpeedDeviation`? By default it's set to a very low value.

This is also mentioned in the EGM manual, and in [abb_egm_hardware_interface/README.md](https://github.com/ros-industrial/abb_robot_driver/blob/master/abb_egm_hardware_interface/README.md#troubleshooting), the *Motion (robot side settings)* row.

If you were using `abb_robot_driver`, you could use the `/rws/get_egm_settings` and `/rws/set_egm_settings` services, discussed in https://github.com/ros-industrial/abb_robot_driver/issues/16#issuecomment-790516237.

From your comment it isn't entirely clear to me whether you're using `abb_libegm` by itself, or as part of `abb_robot_driver`.

**Reply by kappa95** [reply]
No I haven't changed, I left the RAPID code provided by the StateMachine Add-in such it is provided. I have seen on the manual that is limiting the speeds up to 1 deg/s for EGM Joint and Poses, tomorrow I will try to modify it! 

Regarding the libraries I am using "both". In the sense that the C++ code picked as example: (for joint trajectories), I have compiled it calling the `abb_libegm` such as in the CMakeList provided in the folder. And I have used the launcher `roslaunch abb_robot_bringup_examples ex2_rws_and_egm_6axis_robot.launch robot_ip:=192.168.125.1` from the `abb_robot_bringup_examples` in the `abb_robot_driver` library for starting the EGM joint service. (Probably I misunderstood how `abb_libegm` and `abb_robot_driver` work and I did a mess).

**Reply by kappa95** [reply]
Thank you so much, Changing directly the RAPID variable `DEFAULT_MAX_SPEED_DEVIATION` in the `initializeEGM` procedure it moves.

**Comment by kappa95** [comment]
Dear community,

Regarding this topic, I have a question on how to control the speed of the motion given a smoother path. 
The goal I want to achieve is that: given a smooth path of poses (with a certain discretization) and a target speed that i want to keep constant along the path which are the settings to use/method to compute them in order to have that speed. 

Also I would like to understand better the methods: `set_duration` etc... for the object: `abb::egm::wrapper::trajectory::PointGoal`.
I have read the manuals about EGM application and the State-Machine Add-in but I don't find any hint.
I have seen in the library that EGM interpolation and I have noticed that it uses a spline of 5th order but no idea on how it has been used.

Thank you so much in advance
